internal/controller/multiclusterservice_controller.go

```diff
@@ -29,10 +37,132 @@ type MultiClusterServiceReconciler struct {
+		return r.reconcileDelete(ctx, mcsvc)
+	}
+
+	if ok := controllerutil.AddFinalizer(mcsvc, hmc.MultiClusterServiceFinalizer); ok {
+		if err := r.Client.Update(ctx, mcsvc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update MultiClusterService %s with finalizer %s: %w", mcsvc.Name, hmc.MultiClusterServiceFinalizer, err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// By using DefaultSystemNamespace we are enforcing that MultiClusterService
+	// may only use ServiceTemplates that are present in the hmc-system namespace.
+	opts, err := HelmChartOpts(ctx, r.Client, l, utils.DefaultSystemNamespace, mcsvc.Spec.Services)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := sveltos.ReconcileClusterProfile(ctx, r.Client, l, mcsvc.Name,
+		sveltos.ReconcileProfileOpts{
+			OwnerReference: &metav1.OwnerReference{
+				APIVersion: hmc.GroupVersion.String(),
+				Kind:       hmc.MultiClusterServiceKind,
+				Name:       mcsvc.Name,
+				UID:        mcsvc.UID,
+			},
+			LabelSelector:  mcsvc.Spec.ClusterSelector,
+			HelmChartOpts:  opts,
+			Priority:       mcsvc.Spec.Priority,
+			StopOnConflict: mcsvc.Spec.StopOnConflict,
+		}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile ClusterProfile: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// HelmChartOpts returns slice of helm chart options to use with Sveltos.
+// Namespace is the namespace of the referred templates in services slice.
+func HelmChartOpts(ctx context.Context, c client.Client, l logr.Logger, namespace string, services []hmc.ServiceSpec) ([]sveltos.HelmChartOpts, error) {
+	opts := []sveltos.HelmChartOpts{}
+
+	// NOTE: The Profile/ClusterProfile object will be updated with
+	// no helm charts if len(mc.Spec.Services) == 0. This will result
+	// in the helm charts being uninstalled on matching clusters if
+	// Profile/ClusterProfile originally had len(m.Spec.Sevices) > 0.
+	for _, svc := range services {
+		if svc.Disable {
+			l.Info(fmt.Sprintf("Skip adding Template %s because Disable=true", svc.Template))
+			continue
+		}
+
+		tmpl := &hmc.ServiceTemplate{}
+		// Here we can use the same namespace for all services
+		// because if the services slice is part of:
+		// 1. ManagedCluster: Then the referred template must be in its own namespace.
+		// 2. MultiClusterService: Then the referred template must be in hmc-system namespace.
+		ref := types.NamespacedName{Name: svc.Template, Namespace: namespace}
+		if err := c.Get(ctx, ref, tmpl); err != nil {
+			return nil, fmt.Errorf("failed to get Template %s: %w", ref.String(), err)
+		}
+
+		source, err := TemplateSource(ctx, c, tmpl)
+		if err != nil {
+			return nil, fmt.Errorf("could not get repository url: %w", err)
+		}
+
+		opts = append(opts, sveltos.HelmChartOpts{
+			Values:        svc.Values,
+			RepositoryURL: source.Spec.URL,
+			// We don't have repository name so chart name becomes repository name.
+			RepositoryName: tmpl.Spec.Helm.ChartName,
+			ChartName: func() string {
```

Comment:
> can we use plain variables or non-anon functions instead of closures here and in the releaseNamespace field?

I think this is easier to understand no? Is computational cost the reason to remove it?

> and a tiny nit (applicable elsewhere where it makes sense): pkg-imports-wise types.NamespacedName -> client.ObjectKey since we already have imported the client pkg

ack